include/nbl/builtin/hlsl/matrix_utils/transformation_matrix_utils.hlsl

```diff
@@ -0,0 +1,197 @@
+#ifndef _NBL_BUILTIN_HLSL_MATRIX_UTILS_TRANSFORMATION_MATRIX_UTILS_INCLUDED_
+#define _NBL_BUILTIN_HLSL_MATRIX_UTILS_TRANSFORMATION_MATRIX_UTILS_INCLUDED_
+#include <nbl/builtin/hlsl/math/quaternion/quaternion.hlsl>
+// TODO: remove this header when deleting vectorSIMDf.hlsl
+#ifndef __HLSL_VERSION
+#include <nbl/core/math/glslFunctions.h>
+#include "vectorSIMD.h"
+#endif
+#include <nbl/builtin/hlsl/matrix_utils/matrix_traits.hlsl>
+#include "nbl/builtin/hlsl/cpp_compat/unroll.hlsl"
+
+namespace nbl
+{
+namespace hlsl
+{
+
+template<typename MatT>
+MatT diagonal(float diagonal = 1)
+{
+	MatT output;
+
+	NBL_UNROLL_LIMITED(4)
+	for (uint32_t i = 0; i < matrix_traits<MatT>::RowCount; ++i)
+		NBL_UNROLL_LIMITED(4)
+		for (uint32_t j = 0; j < matrix_traits<MatT>::ColumnCount; ++j)
+			output[i][j] = 0;
+
+	NBL_UNROLL_LIMITED(4)
+	for (uint32_t diag = 0; diag < matrix_traits<MatT>::RowCount; ++diag)
+		output[diag][diag] = diagonal;
+
+	return output;
+}
+
+template<typename MatT>
+MatT identity()
+{
+	// TODO
+	// static_assert(MatT::Square);
+	return diagonal<MatT>(1);
+}
+
+template<typename T>
+inline matrix<T, 4, 4> getMatrix3x4As4x4(NBL_CONST_REF_ARG(matrix<T, 3, 4>) mat)
+{
+	matrix<T, 4, 4> output;
+	for (int i = 0; i < 3; ++i)
+		output[i] = mat[i];
+	output[3] = float32_t4(0.0f, 0.0f, 0.0f, 1.0f);
+
+	return output;
+}
+
+template<typename T>
+inline matrix<T, 3, 4> extractSub3x4From4x4Matrix(NBL_CONST_REF_ARG(matrix<T, 4, 4>) mat)
+{
+	matrix<T, 3, 4> output;
+	for (int i = 0; i < 3; ++i)
+		output[i] = mat[i];
+
+	return output;
+}
+
+template<typename T, int N>
+inline matrix<T, 3, 3> getSub3x3(NBL_CONST_REF_ARG(matrix<T, N, 4>) mat)
+{
+	return matrix<T, 3, 3>(mat);
+}
+
+template<int N, int M>
+inline matrix<float64_t, N, M> getAs64BitPrecisionMatrix(NBL_CONST_REF_ARG(matrix<float32_t, N, M>) mat)
+{
+	matrix<float64_t, N, M> output;
+	for (int i = 0; i < N; ++i)
+		output[i] = mat[i];
+
+	return output;
+}
```

Comment:
this should be a `_static_cast_helper` spec - cast equal rank matrices to each other with different types